src/plugins/clips-executive/clips/execmon/stalled-sequential-plan.clp

```diff
@@ -0,0 +1,40 @@
+
+;---------------------------------------------------------------------------
+;  stalled-sequential-plan.clp -
+;  detect stalled sequential strictly progressing plans
+;
+;  Created: Wed Feb 06 11:25:16 2019 +0100
+;  Copyright  2012-2019  Tim Niemueller [www.niemueller.de]
+;  Licensed under GPLv2+ license, cf. LICENSE file
+;---------------------------------------------------------------------------
+
+(defrule xm-stalled-sequential-plan-next-action-not-executable
+  "This rule detects sequential plans which have stalled. It applies the
+   strict assumption that a plan will always immediately advance, that is,
+   once an action has finished, the next one is immediately started.
+   This is in line with PDDL plan execution semantics, where actions fully
+   describe the development of predicates. This assumption is voilated,
+   for example, if the execution should wait for preconditions to be
+   fulfilled, e.g., in the presence of sensed predicates which may change
+   due to exogenous events. In this case, a more lenient or specific
+   XM mode may be desirable."
+  (declare (salience ?*SALIENCE-LOW*))
+  ?g <- (goal (id ?goal-id) (mode DISPATCHED) (committed-to ?plan-id))
+  (plan (id ?plan-id) (goal-id ?goal-id) (type SEQUENTIAL))
+
+  ; We do have at least one action which is still to be executed
+  (plan-action (goal-id ?goal-id) (plan-id ?plan-id) (id ?id)
+               (state FORMULATED) (executable FALSE))
+
+  ; There is no other FORMULATED action which comes earlier
+  (not (plan-action (goal-id ?goal-id) (plan-id ?plan-id)
+                    (state FORMULATED) (id ?oid&:(< ?oid ?id))))
+
+  ; there is no action which is currently ongoing
+  (not (plan-action (goal-id ?goal-id) (plan-id ?plan-id)
+                    (state ~FORMULATED&~FINAL&~FAILED)))
+ =>
+  (modify ?g (mode FINISHED) (outcome FAILED)
+             (error STALLED-NONE-EXECUTABLE)
+             (message "No action of the plan was executable"))
```

Comment:
It would make sense to specifically log the action that stalled (`?id` in this case)

Reply:
I thought about that, but without having a reference of the plan at hand (printing the full plan) it doesn't help much. However, the action-name often already helps far enough. When using webview you can see where the plan stalled right away anyway (when disabling auto-retraction that is).

Reply:
But the action name does not occur in the message either? I think action ID + name (and maybe the plan ID) would help a lot, adding the parameters to the message would also be possible. Currently, all the message conveys is that the plan could not be continued, but we know more than that, we know which action stalled.

Reply:
Oh, indeed. Let me do something about that. I had this reasoning in another case (action-selection-sequential-failed) where I decided not to print the ID and that came up in my head when thinking about it.

Reply:
Just pushed the most verbose we can do in the case (short of printing the plan that is).